File: android/src/main/kotlin/com/telemetrydeck/telemetrydecksdk/TelemetrydecksdkPlugin.kt
package com.telemetrydeck.telemetrydecksdk

import android.app.Application
import android.content.Context
import com.telemetrydeck.sdk.PurchaseEvent
import com.telemetrydeck.sdk.PurchaseType
import com.telemetrydeck.sdk.TelemetryDeck
import com.telemetrydeck.sdk.params.ErrorCategory
import com.telemetrydeck.sdk.providers.DefaultParameterProvider
import com.telemetrydeck.sdk.providers.DefaultPrefixProvider
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** TelemetrydecksdkPlugin */
class TelemetrydecksdkPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private var applicationContext: Context? = null
    private val coroutineScope =
        CoroutineScope(Dispatchers.IO) // Coroutine scope for background tasks

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "telemetrydecksdk")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "start" -> {
                nativeInitialize(call, result)
            }

            "stop" -> {
                nativeStop(result)
            }

            "send" -> {
                // this maps to the queue method which aligns with the behaviour of the iOS SDK
                nativeQueue(call, result)
            }

            "startDurationSignal" -> {
                // this maps to the queue method which aligns with the behaviour of the iOS SDK
                nativeStartDurationSignal(call, result)
            }

            "stopAndSendDurationSignal" -> {
                // this maps to the queue method which aligns with the behaviour of the iOS SDK
                nativeStopAndSendDurationSignal(call, result)
            }

            "generateNewSession" -> {
                TelemetryDeck.newSession()
                result.success(null)
            }

            "updateDefaultUser" -> {
                nativeUpdateDefaultUser(call, result)
            }

            "navigate" -> {
                nativeNavigate(call, result)
            }

            "navigateToDestination" -> {
                nativeNavigateDestination(call, result)
            }

            "acquiredUser" -> {
                nativeAcquiredUser(call, result)
            }

            "leadStarted" -> {
                nativeLeadStarted(call, result)
            }

            "leadConverted" -> {
                nativeLeadConverted(call, result)
            }

            "onboardingCompleted" -> {
                nativeOnboardingCompleted(call, result)
            }

            "coreFeatureUsed" -> {
                nativeCoreFeatureUsed(call, result)
            }

            "paywallShown" -> {
                nativePaywallShown(call, result)
            }

            "purchaseCompleted" -> {
                nativePurchaseCompleted(call, result)
            }

            "referralSent" -> {
                nativeReferralSent(call, result)
            }

            "userRatingSubmitted" -> {
                nativeUserRatingSubmitted(call, result)
            }

            "errorOccurred" -> {
                nativeErrorOccurred(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    /**
     * Send a signal that represents a navigation event with a source and a destination.
     *
     * @see <a href="https://telemetrydeck.com/docs/articles/navigation-signals/">Navigation Signals</a>
     * */
    private fun nativeNavigate(call: MethodCall, result: Result) {
        val sourcePath = call.argument<String>("sourcePath")
        val destinationPath = call.argument<String>("destinationPath")
        val clientUser = call.argument<String?>("clientUser")

        if (sourcePath == null || destinationPath == null) {
            result.error("INVALID_ARGUMENT", "sourcePath and destinationPath are required", null)
            return
        }

        coroutineScope.launch {
            TelemetryDeck.navigate(sourcePath, destinationPath, clientUser)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    /**
     * Send a signal that represents a navigation event with a destination and a default source.
     *
     * @see <a href="https://telemetrydeck.com/docs/articles/navigation-signals/">Navigation Signals</a>
     * */
    private fun nativeNavigateDestination(call: MethodCall, result: Result) {
        val destinationPath = call.argument<String>("destinationPath")
        val clientUser = call.argument<String?>("clientUser")

        if (destinationPath == null) {
            result.error("INVALID_ARGUMENT", "destinationPath is required", null)
            return
        }

        coroutineScope.launch {
            TelemetryDeck.navigate(destinationPath, clientUser)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeStop(result: Result) {
        TelemetryDeck.stop()
        result.success(null)
    }

    private fun nativeUpdateDefaultUser(
        call: MethodCall,
        result: Result
    ) {
        val user = call.arguments<String>()

        coroutineScope.launch {
            TelemetryDeck.newDefaultUser(user)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeQueue(
        call: MethodCall,
        result: Result
    ) {
        val signalType = call.argument<String>("signalType")
        if (signalType != null) {
            val clientUser = call.argument<String?>("clientUser")
            val additionalPayload = call.argument<Map<String, String>>("additionalPayload")
            val floatValue = call.argument<Double?>("floatValue")

            coroutineScope.launch {
                TelemetryDeck.signal(signalType, additionalPayload.orEmpty(), floatValue, clientUser)

                withContext(Dispatchers.Main) {
                    result.success(null)
                }
            }
        } else {
            result.error("INVALID_ARGUMENT", "signalType must be provided", null)
        }
    }

    private fun nativeStartDurationSignal(
        call: MethodCall,
        result: Result
    ) {
        val signalType = call.argument<String>("signalType")
        if (signalType != null) {
            val parameters = call.argument<Map<String, String>>("parameters")

            coroutineScope.launch {
                TelemetryDeck.startDurationSignal(signalType, parameters.orEmpty())

                withContext(Dispatchers.Main) {
                    result.success(null)
                }
            }
        } else {
            result.error("INVALID_ARGUMENT", "signalType must be provided", null)
        }
    }

    private fun nativeStopAndSendDurationSignal(
        call: MethodCall,
        result: Result
    ) {
        val signalType = call.argument<String>("signalType")
        if (signalType != null) {
            val parameters = call.argument<Map<String, String>>("parameters")

            coroutineScope.launch {
                TelemetryDeck.stopAndSendDurationSignal(signalType, parameters.orEmpty())

                withContext(Dispatchers.Main) {
                    result.success(null)
                }
            }
        } else {
            result.error("INVALID_ARGUMENT", "signalType must be provided", null)
        }
    }

    private fun nativeAcquiredUser(call: MethodCall, result: Result) {
        val channel = call.argument<String>("channel")
        if (channel == null) {
            result.error("INVALID_ARGUMENT", "channel is required", null)
            return
        }

        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.acquiredUser(channel, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeLeadStarted(call: MethodCall, result: Result) {
        val leadId = call.argument<String>("leadId")
        if (leadId == null) {
            result.error("INVALID_ARGUMENT", "leadId is required", null)
            return
        }

        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.leadStarted(leadId, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeLeadConverted(call: MethodCall, result: Result) {
        val leadId = call.argument<String>("leadId")
        if (leadId == null) {
            result.error("INVALID_ARGUMENT", "leadId is required", null)
            return
        }

        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.leadConverted(leadId, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeOnboardingCompleted(call: MethodCall, result: Result) {
        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.onboardingCompleted(params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeCoreFeatureUsed(call: MethodCall, result: Result) {
        val featureName = call.argument<String>("featureName")
        if (featureName == null) {
            result.error("INVALID_ARGUMENT", "featureName is required", null)
            return
        }

        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.coreFeatureUsed(featureName, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativePaywallShown(call: MethodCall, result: Result) {
        val reason = call.argument<String>("reason")
        if (reason == null) {
            result.error("INVALID_ARGUMENT", "reason is required", null)
            return
        }

        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.paywallShown(reason, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativePurchaseCompleted(call: MethodCall, result: Result) {
        val eventString = call.argument<String>("event")
        val countryCode = call.argument<String>("countryCode")
        val productID = call.argument<String>("productID")
        val purchaseTypeString = call.argument<String>("purchaseType")
        val priceAmountMicros = call.argument<Long>("priceAmountMicros")
        val currencyCode = call.argument<String>("currencyCode")

        if (eventString == null || countryCode == null || productID == null ||
            purchaseTypeString == null || priceAmountMicros == null || currencyCode == null) {
            result.error("INVALID_ARGUMENT", "event, countryCode, productID, purchaseType, priceAmountMicros, and currencyCode are required", null)
            return
        }

        val event = when (eventString) {
            "purchaseCompleted" -> PurchaseEvent.PAID_PURCHASE
            "freeTrialStarted" -> PurchaseEvent.STARTED_FREE_TRIAL
            "convertedFromFreeTrial" -> PurchaseEvent.CONVERTED_FROM_TRIAL
            else -> {
                result.error("INVALID_ARGUMENT", "Invalid purchase event: $eventString", null)
                return
            }
        }

        val purchaseType = when (purchaseTypeString) {
            "subscription" -> PurchaseType.SUBSCRIPTION
            "oneTimePurchase" -> PurchaseType.ONE_TIME_PURCHASE
            else -> {
                result.error("INVALID_ARGUMENT", "Invalid purchase type: $purchaseTypeString", null)
                return
            }
        }

        val offerID = call.argument<String?>("offerID")
        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.purchaseCompleted(
                event,
                countryCode,
                productID,
                purchaseType,
                priceAmountMicros,
                currencyCode,
                offerID,
                params.orEmpty(),
                customUserID
            )
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeReferralSent(call: MethodCall, result: Result) {
        val receiversCount = call.argument<Int>("receiversCount")
        if (receiversCount == null) {
            result.error("INVALID_ARGUMENT", "receiversCount is required", null)
            return
        }

        val kind = call.argument<String?>("kind")
        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.referralSent(receiversCount, kind, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeUserRatingSubmitted(call: MethodCall, result: Result) {
        val rating = call.argument<Int>("rating")
        if (rating == null) {
            result.error("INVALID_ARGUMENT", "rating is required", null)
            return
        }

        val comment = call.argument<String?>("comment")
        val params = call.argument<Map<String, String>>("params")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.userRatingSubmitted(rating, comment, params.orEmpty(), customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeErrorOccurred(call: MethodCall, result: Result) {
        val id = call.argument<String>("id")
        if (id == null) {
            result.error("INVALID_ARGUMENT", "id is required", null)
            return
        }

        val categoryString = call.argument<String?>("category")
        val category = when (categoryString) {
            "thrownException" -> ErrorCategory.ThrownException
            "userInput" -> ErrorCategory.UserInput
            "appState" -> ErrorCategory.AppState
            else -> null
        }

        val message = call.argument<String?>("message")
        val parameters = call.argument<Map<String, String>>("parameters")
        val floatValue = call.argument<Double?>("floatValue")
        val customUserID = call.argument<String?>("customUserID")

        coroutineScope.launch {
            TelemetryDeck.errorOccurred(id, category, message, parameters.orEmpty(), floatValue, customUserID)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        }
    }

    private fun nativeInitialize(call: MethodCall, result: Result) {
        val arguments = call.arguments as? Map<*, *> // Cast to a Map
        if (arguments != null) {
            // Extract values using the expected keys
            // we extract the required appID parameter
            val appID = arguments["appID"] as? String
            if (appID == null) {
                result.error("INVALID_ARGUMENT", "Expected value appID is not provided.", null)
                return
            }

            // additional optional parameters
            val apiBaseURL = arguments["apiBaseURL"] as? String?
            val defaultUser = arguments["defaultUser"] as? String?
            val debug = arguments["debug"] as? Boolean
            val testMode = arguments["testMode"] as? Boolean
            val salt = arguments["salt"] as? String?
            val defaultSignalPrefix = arguments["defaultSignalPrefix"] as? String?
            val defaultParameterPrefix = arguments["defaultParameterPrefix"] as? String?
            val defaultParameters = arguments["defaultParameters"] as? Map<String, String>?
            val namespace = arguments["namespace"] as? String?


            // Initialize the client
            // Do not activate the lifecycle provider
            val builder = TelemetryDeck.Builder()
                .appID(appID)

            apiBaseURL?.let {
                builder.baseURL(it)
            }
            defaultUser?.let {
                builder.defaultUser(it)
            }
            debug?.let {
                builder.showDebugLogs(it)
            }
            testMode?.let {
                builder.testMode(it)
            }
            salt?.let {
                builder.salt(it)
            }
            namespace?.let {
                builder.namespace(it)
            }

            if (defaultSignalPrefix != null || defaultParameterPrefix != null) {
                builder.addProvider(DefaultPrefixProvider(defaultSignalPrefix, defaultParameterPrefix))
            }

            if (defaultParameters != null) {
                builder.addProvider(DefaultParameterProvider(defaultParameters))
            }

            val application = applicationContext as Application
            TelemetryDeck.start(application, builder)
            result.success(null)
        } else {
            result.error("INVALID_ARGUMENT", "Arguments are not a map", null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/telemetrydecksdk/Sources/telemetrydecksdk/TelemetrydecksdkPlugin.swift
import Flutter
import UIKit
import TelemetryDeck

public class TelemetrydecksdkPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "telemetrydecksdk", binaryMessenger: registrar.messenger())
        let instance = TelemetrydecksdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            nativeInitialize(call, result: result)
        case "stop":
            nativeStop(call, result: result)
        case "send":
            nativeQueue(call, result: result)
        case "startDurationSignal":
            nativeStartDurationSignal(call, result: result)
        case "stopAndSendDurationSignal":
            nativeStopAndSendDurationSignal(call, result: result)
        case "generateNewSession":
            TelemetryDeck.generateNewSession()
            result(nil)
        case "updateDefaultUser":
            nativeUpdateDefaultUser(call, result: result)
        case "navigate":
            nativeNavigate(call, result: result)
        case "navigateToDestination":
            nativeNavigateDestination(call, result: result)
        case "acquiredUser":
            nativeAcquiredUser(call, result: result)
        case "leadStarted":
            nativeLeadStarted(call, result: result)
        case "leadConverted":
            nativeLeadConverted(call, result: result)
        case "onboardingCompleted":
            nativeOnboardingCompleted(call, result: result)
        case "coreFeatureUsed":
            nativeCoreFeatureUsed(call, result: result)
        case "paywallShown":
            nativePaywallShown(call, result: result)
        case "purchaseCompleted":
            nativePurchaseCompleted(call, result: result)
        case "referralSent":
            nativeReferralSent(call, result: result)
        case "userRatingSubmitted":
            nativeUserRatingSubmitted(call, result: result)
        case "errorOccurred":
            nativeErrorOccurred(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    /**
     * Send a signal that represents a navigation event with a source and a destination.
     *
     * @see <a href="https://telemetrydeck.com/docs/articles/navigation-signals/">Navigation Signals</a>
     * */
    private func nativeNavigate(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let sourcePath = arguments["sourcePath"] as? String,
              let destinationPath = arguments["destinationPath"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "sourcePath and destinationPath are required", details: nil))
            return
        }
        let clientUser = arguments["clientUser"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.navigationPathChanged(from: sourcePath, to: destinationPath, customUserID: clientUser)
            result(nil)
        }
    }
    
    /**
     * Send a signal that represents a navigation event with a destination and a default source.
     *
     * @see <a href="https://telemetrydeck.com/docs/articles/navigation-signals/">Navigation Signals</a>
     * */
    private func nativeNavigateDestination(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let destinationPath = arguments["destinationPath"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "destinationPath are required", details: nil))
            return
        }
        let clientUser = arguments["clientUser"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.navigationPathChanged(to: destinationPath, customUserID: clientUser)
            result(nil)
        }
    }
    
    private func nativeStop(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        TelemetryDeck.terminate()
        result(nil)
    }
    
    private func nativeUpdateDefaultUser(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        TelemetryDeck.updateDefaultUserID(to: call.arguments as? String)
        result(nil)
    }
    
    private func nativeQueue(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any], let signalType = arguments["signalType"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing required argument signalType", details: nil))
            return
        }

        let clientUser = arguments["clientUser"] as? String
        let additionalPayload = arguments["additionalPayload"] as? [String : String] ?? [:]
        let floatValue = arguments["floatValue"] as? Double

        // do not attempt to send signals if the client is stopped
        if TelemetryManager.isInitialized {
            TelemetryDeck.signal(signalType, parameters: additionalPayload, floatValue: floatValue, customUserID: clientUser)
        }

        result(nil)
    }
    
    private func nativeStartDurationSignal(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any], let signalType = arguments["signalType"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing required argument signalType", details: nil))
            return
        }
        
        let parameters = arguments["parameters"] as? [String : String] ?? [:]
        
        // do not attempt to send signals if the client is stopped
        if TelemetryManager.isInitialized {
            DispatchQueue.main.async {
                TelemetryDeck.startDurationSignal(signalType, parameters: parameters)
                result(nil)
            }
        } else {
            result(nil)
        }
    }
    
    private func nativeStopAndSendDurationSignal(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any], let signalType = arguments["signalType"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing required argument signalType", details: nil))
            return
        }
        
        let parameters = arguments["parameters"] as? [String : String] ?? [:]
        
        // do not attempt to send signals if the client is stopped
        if TelemetryManager.isInitialized {
            DispatchQueue.main.async {
                TelemetryDeck.stopAndSendDurationSignal(signalType, parameters: parameters)
                result(nil)
            }
        } else {
            result(nil)
        }
    }
    
    private func nativeInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Arguments are not a map", details: nil))
            return
        }
        
        // appD is required
        guard let appID: String = arguments["appID"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected value appID is not provided.", details: nil))
            return
        }
        
        // baseURL is optional but part of the `TelemetryManagerConfiguration` initializer
        var baseURL: URL? = nil
        if let apiBaseURL = arguments["apiBaseURL"] as? String, let url = URL(string: apiBaseURL) {
            baseURL = url
        }
        
        
        let salt =  arguments["salt"] as? String
        let namespace = arguments["namespace"] as? String
        let configuration = TelemetryManagerConfiguration.init(
            appID: appID, salt: salt, baseURL: baseURL, namespace: namespace)
        
        // other optional params
        if arguments.keys.contains("defaultUser") {
            configuration.defaultUser = arguments["defaultUser"] as? String
        }
        
        if arguments.keys.contains("debug") {
            if arguments["debug"] as? Bool == true {
                // by default, the library logs with level .info
                configuration.logHandler = LogHandler.standard(.debug)
            }
        }
        
        if arguments.keys.contains("testMode") {
            configuration.testMode = arguments["testMode"] as? Bool == true
        }
        
        if arguments.keys.contains("defaultSignalPrefix") {
            configuration.defaultSignalPrefix = arguments["defaultSignalPrefix"] as? String
        }
        
        if arguments.keys.contains("defaultParameterPrefix") {
            configuration.defaultParameterPrefix = arguments["defaultParameterPrefix"] as? String
        }
        
        if arguments.keys.contains("defaultParameters") {
            let finalValues = arguments["defaultParameters"] as? [String : String] ?? [:]
            configuration.defaultParameters = { finalValues }
        }
        
        TelemetryDeck.initialize(config: configuration)

        result(nil)
    }

    private func nativeAcquiredUser(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let channel = arguments["channel"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "channel is required", details: nil))
            return
        }
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.acquiredUser(channel: channel, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativeLeadStarted(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let leadId = arguments["leadId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "leadId is required", details: nil))
            return
        }
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.leadStarted(leadID: leadId, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativeLeadConverted(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let leadId = arguments["leadId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "leadId is required", details: nil))
            return
        }
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.leadConverted(leadID: leadId, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativeOnboardingCompleted(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Arguments are not a map", details: nil))
            return
        }
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.onboardingCompleted(parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativeCoreFeatureUsed(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let featureName = arguments["featureName"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "featureName is required", details: nil))
            return
        }
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.coreFeatureUsed(featureName: featureName, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativePaywallShown(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let reason = arguments["reason"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "reason is required", details: nil))
            return
        }
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.paywallShown(reason: reason, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativePurchaseCompleted(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let eventString = arguments["event"] as? String,
              let countryCode = arguments["countryCode"] as? String,
              let productID = arguments["productID"] as? String,
              let purchaseTypeString = arguments["purchaseType"] as? String,
              let priceAmountMicros = arguments["priceAmountMicros"] as? Int,
              let currencyCode = arguments["currencyCode"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "event, countryCode, productID, purchaseType, priceAmountMicros, and currencyCode are required", details: nil))
            return
        }

        let offerID = arguments["offerID"] as? String
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String

        let signalName: String
        switch eventString {
        case "freeTrialStarted":
            signalName = "TelemetryDeck.Purchase.freeTrialStarted"
        case "convertedFromFreeTrial":
            signalName = "TelemetryDeck.Purchase.convertedFromFreeTrial"
        default:
            signalName = "TelemetryDeck.Purchase.completed"
        }

        let type: String
        switch purchaseTypeString {
        case "subscription":
            type = "subscription"
        default:
            type = "one-time-purchase"
        }

        var purchaseParams: [String: String] = [
            "TelemetryDeck.Purchase.type": type,
            "TelemetryDeck.Purchase.countryCode": countryCode,
            "TelemetryDeck.Purchase.productID": productID,
            "TelemetryDeck.Purchase.currencyCode": currencyCode
        ]

        if let offerID = offerID {
            purchaseParams["TelemetryDeck.Purchase.offerID"] = offerID
        }

        let mergedParams = purchaseParams.merging(params) { $1 }

        let priceInUSD = Double(priceAmountMicros) / 1_000_000.0

        DispatchQueue.main.async {
            TelemetryDeck.signal(
                signalName,
                parameters: mergedParams,
                floatValue: priceInUSD,
                customUserID: customUserID
            )
            result(nil)
        }
    }

    private func nativeReferralSent(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Arguments are not a map", details: nil))
            return
        }
        let receiversCount = arguments["receiversCount"] as? Int ?? 1
        let kind = arguments["kind"] as? String
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.referralSent(receiversCount: receiversCount, kind: kind, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativeUserRatingSubmitted(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let rating = arguments["rating"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "rating is required", details: nil))
            return
        }
        let comment = arguments["comment"] as? String
        let params = arguments["params"] as? [String: String] ?? [:]
        let customUserID = arguments["customUserID"] as? String
        DispatchQueue.main.async {
            TelemetryDeck.userRatingSubmitted(rating: rating, comment: comment, parameters: params, customUserID: customUserID)
            result(nil)
        }
    }

    private func nativeErrorOccurred(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let id = arguments["id"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "id is required", details: nil))
            return
        }

        let categoryString = arguments["category"] as? String
        let category: ErrorCategory?
        if let categoryString = categoryString {
            switch categoryString {
            case "thrownException":
                category = .thrownException
            case "userInput":
                category = .userInput
            case "appState":
                category = .appState
            default:
                category = nil
            }
        } else {
            category = nil
        }

        let message = arguments["message"] as? String
        let parameters = arguments["parameters"] as? [String: String] ?? [:]
        let floatValue = arguments["floatValue"] as? Double
        let customUserID = arguments["customUserID"] as? String

        DispatchQueue.main.async {
            TelemetryDeck.errorOccurred(
                id: id,
                category: category,
                message: message,
                parameters: parameters,
                floatValue: floatValue,
                customUserID: customUserID
            )
            result(nil)
        }
    }
}
